File: scripts/prepare-bower.pl
#!/usr/bin/env perl

use warnings;
use strict;
use feature 'say';

=head1 prepare-bower.pl

This is a script for preparing a snippet for adding to a Dhall file of Spacchetti package definitions.

See L<the Spachetti repository|https://github.com/spacchetti/spacchetti> for more details.

Usage:

  prepare-bower.pl [package-name]

Example:

  ./prepare-bower.pl yargs

  => yargs = mkPackage
     [
       "console",
       "either",
       "exceptions",
       "foreign",
       "unsafe-coerce"
     ]
     "https://github.com/paf31/purescript-yargs.git"
     "4.0.0"

=cut

my $numArgs = $#ARGV;

if ($numArgs < 0) {
  print "I need one arg for what the bower package name is without the preceding `purescript-`\n";
  print "e.g. `./scripts/prepare-bower.pl yargs`\n";
  exit;
}

my $input = $ARGV[0];
my $json = "bower-info/$input.json";

unless(-e $json) {
    print `mkdir -p bower-info`;
    print `bower info purescript-$input --json > $json`;
}

chomp(my $dependenciesCheck = `cat $json | jq '.latest.dependencies?'`);
my $dependencies = '';

unless ($dependenciesCheck eq 'null') {
  chomp($dependencies = `cat $json | jq '.latest.dependencies' | jq 'keys | map(.[11:])'`);
}

chomp(my $version = `cat $json | jq '.latest.version'`);
chomp(my $url = `cat $json | jq '.latest.repository.url'`);
if ($url eq 'null') {
  # naively use homepage if url isn't specified for repo
  chomp($url = `cat $json | jq '.latest.homepage'`);
  $url =~ s/"//g;
  $url .= '.git';
  $url = '"' . $url . '"';
} else {
  $url =~ s/git:/https:/;
  $url =~ s/com:/com\//;
  $url =~ s/git@/https:\/\//g;
  unless ($url =~ /\.git"$/) {
    chop $url;
    $url .= '.git"';
  }
}

# bower doesn't believe in telling us the correct tag names
unless ($version =~ /v/) {
    $version =~ s/"//g;
    $version = "\"v$version\"";
}

my $output = <<END_TEMPLATE;
$input = mkPackage
$dependencies
$url
$version
END_TEMPLATE

print $output

File: scripts/prepare-bower.hs
#!/usr/bin/env runhaskell

{-# LANGUAGE OverloadedStrings #-}

import Control.Monad
import qualified Data.List as List
import qualified Data.Text as Text
import qualified System.Environment as Env
import qualified System.Exit as Exit
import qualified System.Directory as Dir
import qualified System.Process as Proc

overText :: (Text.Text -> Text.Text) -> String -> String
overText f = Text.unpack . f. Text.pack

strip :: String -> String
strip = overText Text.strip

cleanupUrl :: String -> String
cleanupUrl s = "\"" <> prepare s <> ".git\""
  where
    prepare
      = overText
      $ Text.replace "git:" "https:"
      . Text.replace "com:" "com\\/"
      . Text.replace "git@" "https:\\/\\/"
      . Text.replace ".git" ""
      . Text.replace "\"" ""

-- bower doesn't believe in telling us the correct tag names
cleanupVersion :: String -> String
cleanupVersion version
  | not $ List.elem 'v' version
  , withoutQuotes <- overText (Text.replace "\"" "") version
  = "\"v" <> withoutQuotes <> "\""
cleanupVersion version = version

run :: String -> IO ()
run input' = do
  test <- Dir.doesFileExist file

  when (not test) $ void $ Proc.readCreateProcess downloadJsonCmd ""

  (exitCode, depsOut, _) <- Proc.readCreateProcessWithExitCode getDeps ""
  dependencies <- case exitCode of
    Exit.ExitSuccess -> pure $ strip depsOut
    Exit.ExitFailure _ -> pure ""

  version <- cleanupVersion <$> readCmd getVersion
  url' <- readCmd getUrl
  url <- if url' == "null"
    then do
      -- naively use homepage if url isn't specified for repo
      output <- readCmd getHomepage
      pure $ "\"" <> output <> ".git" <> "\""
    else
      pure $ cleanupUrl url'

  putStrLn $ input <> " = mkPackage"
  putStrLn dependencies
  putStrLn url
  putStrLn version

  where
    readCmd cmd = strip <$> Proc.readCreateProcess cmd ""

    input = List.filter ((/=) '"') input'
    file = "bower-info/" <> input <> ".json"

    replaceFile = Text.replace "$file" (Text.pack file)
    replaceInput = Text.replace "$input" (Text.pack input)
    mkShell = Proc.shell . overText (replaceInput . replaceFile)

    downloadJsonCmd = mkShell "mkdir -p bower-info && bower info purescript-$input --json > $file"
    getDeps = mkShell "cat $file | jq '.latest.dependencies | keys | map (.[11:])'"
    getVersion = mkShell "cat $file | jq '.latest.version'"
    getUrl = mkShell "cat $file | jq '.latest.repository.url'"
    getHomepage = mkShell "cat $file | jq '.latest.homepage' -r"

main :: IO ()
main = do
  args <- Env.getArgs
  case args of
    [pkg] -> run pkg
    _ -> fail errorMsg

errorMsg :: String
errorMsg = "\n\
\  I need one arg for what the bower package name is without the preceding `purescript-`\n\
\  e.g. `./scripts/prepare-bower.hs yargs`\n"
